files/nrpe/check_openstack.py

```diff
@@ -672,6 +681,53 @@ def execute(self):
     except:
       raise
 
+class OSL3Agent():
+  '''
+  Check all Neutron L3 agents are alive
+  '''
+
+  options = dict()
+
+  def __init__(self, options):
+    self.neutron = neutronclient.Client('2', session=keystone_session_v3(options))
+
+  def check_bad_l3_agents(self):
+    err_l3agents = []
+    l3agents = self.neutron.list_agents(agent_type='L3 agent')
+
+    if str(l3agents.keys()) == "[u'agents']":
```

Comment:
Here it might be not knowing exactly the structure returned by list_agents. But, are you checking that there is a key called 'agents'? u'agents' will work while the function return keys in Unicode (old Python2) code, but with Python3 it would be a string that defualts to a byte structure. If you are checking that there is a key called agents, the best way problaby is to check if 'agents' in l3agents, and then for agent in l3agents['agents'] (if the response is similar to what this doc shows https://docs.openstack.org/api-ref/network/v2/index.html?expanded=list-all-agents-detail#agents)

Reply:
Fixed